### PR TITLE
`WithTenantId` on CIAM authorities

### LINKED ISSUE #4191: [Bug] AbstractAcquireTokenParameterBuilder<T>.WithTenantId() should be supported for CIAM authorities
When doing an OBO, if no tenant is specified, it should be possible to set the tenant to the user tenant (from the tid claim). This will come as a GUID, whereas the authority, in the case of CIAM, is a domain name.

This is blocking Microsoft.Identity.Web OBO samples for CIAM,.

**Logs and network traces**
>	Microsoft.Identity.Client.dll!Microsoft.Identity.Client.AbstractAcquireTokenParameterBuilder<Microsoft.Identity.Client.AcquireTokenOnBehalfOfParameterBuilder>.WithTenantId(string tenantId) Line 272	C#
 	Microsoft.Identity.Web.TokenAcquisition.dll!Microsoft.Identity.Web.TokenAcquisition.GetAuthenticationResultForWebApiToCallDownstreamApiAsync(Microsoft.Identity.Client.IConfidentialClientApplication application, string tenantId, System.Collections.Generic.IEnumerable<string> scopes, Microsoft.Identity.Web.TokenAcquisitionOptions tokenAcquisitionOptions, Microsoft.Identity.Web.MergedOptions mergedOptions) Line 727	C#
 	Microsoft.Identity.Web.TokenAcquisition.dll!Microsoft.Identity.Web.TokenAcquisition.GetAuthenticationResultForUserAsync(System.Collections.Generic.IEnumerable<string> scopes, string authenticationScheme, string tenantId, string userFlow, System.Security.Claims.ClaimsPrincipal user, Microsoft.Identity.Web.TokenAcquisitionOptions tokenAcquisitionOptions) Line 243	C#
 	Microsoft.Identity.Web.TokenAcquisition.dll!Microsoft.Identity.Web.TokenAcquisition.GetAccessTokenForUserAsync(System.Collections.Generic.IEnumerable<string> scopes, string authenticationScheme, string tenantId, string userFlow, System.Security.Claims.ClaimsPrincipal user, Microsoft.Identity.Web.TokenAcquisitionOptions tokenAcquisitionOptions) Line 500	C#
 	Microsoft.Identity.Web.TokenAcquisition.dll!Microsoft.Identity.Web.ITokenAcquisition.GetAccessTokenForUserAsync(System.Collections.Generic.IEnumerable<string> scopes, string tenantId, string userFlow, System.Security.Claims.ClaimsPrincipal user, Microsoft.Identity.Web.TokenAcquisitionOptions tokenAcquisitionOptions) Line 42	C#
 	TodoListService.dll!TodoListService.Controllers.TodoListController.GetAsync() Line 57	C#

**Which version of MSAL.NET are you using?**
4.54.1

**What authentication flow has the issue?**
    * [ x] On-Behalf-Of

Other?
With a CIAM authority

**Is this a new or existing app?**

```csharp
     ClaimsPrincipal? user = _tokenAcquisitionHost.GetUserFromRequest();
      var userTenant = string.Empty;
      if (user != null)
      {
          userTenant = user.GetTenantId();
          builder.WithCcsRoutingHint(user.GetObjectId(), userTenant);
      }
      if (!string.IsNullOrEmpty(tenantId))
      {
          builder.WithTenantId(tenantId);
      }
      else
      {
          if (!string.IsNullOrEmpty(userTenant))
          {
              builder.WithTenantId(userTenant);
          }
      }
```

**Actual behavior**
Exception:
`Microsoft.Identity.Client.MsalClientException: 'WithTenantId can only be used when an AAD authority is specified at the application level.'`

**Expected behavior**
It should be possible to override the tenant with a CIAM authority.
It's not up to MSAL.NET to decide if the IdP will reject it or not (it won't in that case)

**Possible solution**
`AuthorityInfo.IsTenantOverrideSupported` shoud be set to true for CIAM authority

**Additional context / logs / screenshots / links to code**

### PR DESCRIPTION
**Pull request by bgavrilMS**
Fixes #4191

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs [hunk 1]
```diff
@@ -140,16 +140,13 @@ internal class AuthorityInfo
             AuthorityType != AuthorityType.Adfs && 
             AuthorityType != AuthorityType.B2C;
 
-        internal bool IsTenantOverrideSupported =>
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:143]: No need for a flag, but an implementation of how to perform tenant override.

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs [hunk 2]
```diff
@@ -62,25 +83,130 @@ public void WithTenantIdExceptions()
...
+            Assert.AreEqual(
+                TestConstants.TenantId2, 
+                AuthorityHelpers.GetTenantId(new Uri(cca.Authority)));
+        }
+
+        [DataTestMethod]
+        [DataRow(TestConstants.ADFSAuthority, null)]
+        [DataRow(TestConstants.B2CAuthority, "tenant")]
+        [DataRow(TestConstants.AuthorityGuidTenant, TestConstants.TenantIdNumber1)]
+        [DataRow(TestConstants.DstsAuthorityTenanted, TestConstants.TenantId)]
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:149]: I thought the agreement was `WithTenantId` overwrites the authority every time.

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs [hunk 3]
```diff
@@ -62,25 +83,130 @@ public void WithTenantIdExceptions()
...
+                (cca.AppConfig as ApplicationConfiguration).Authority.AuthorityInfo.Host,
+                "The host should have stayed the same");
+
+            Assert.AreEqual(
+               expectedTenant,
+               AuthorityHelpers.GetTenantId(new Uri(cca.Authority)));
+        }
+
+        [TestMethod]
+        public void GenericAuthorityWithTenant()
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:169]: Same as above comment, should match GenericAuthority_WithTenantId_Exceptions?

File: src/client/Microsoft.Identity.Client/Instance/Authority.cs [hunk 4]
```diff
@@ -106,10 +106,17 @@ internal static Authority CreateAuthorityWithEnvironment(AuthorityInfo authority
...
 
         /// <summary>
-        /// Gets a tenanted authority if the current authority is tenant-less.
-        /// Returns the original authority on B2C and ADFS
+        /// Changes the tenant id of the authority, if the authority supports tenants. If not, throws exception,.
         /// </summary>
-        internal abstract string GetTenantedAuthority(string tenantId, bool forceTenantless = false);
+        /// <param name="tenantId">The new tenant id</param>
+        /// <param name="forceTenantless">Forces the change, even if the current tenant is not "common" or "organizations" or "consumers"</param>
+        internal virtual string GetTenantedAuthority(string tenantId, bool forceTenantless)
```
[the remark this answers is not in the document]
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Instance/Authority.cs:113]: Renamed. I'm not going to add a default param to force MSAL developers to read the docs on what this flag does.

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs [hunk 5]
```diff
@@ -62,25 +83,130 @@ public void WithTenantIdExceptions()
...
                .WithClientSecret("secret")
                .Build();
 
-            var ex2 = AssertException.Throws<MsalClientException>(() =>
-                app2
-                    .AcquireTokenByAuthorizationCode(TestConstants.s_scope, "code")
-                    .WithTenantId(TestConstants.TenantId));
+            var parameterBuilder = app
+                .AcquireTokenByAuthorizationCode(TestConstants.s_scope, "code")
+                .WithTenantId(TestConstants.TenantId);
```
[the remark this answers is not in the document]
Comment (bgavrilMS) [reply tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:105]: Good catch!

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs (hunk 2 above)
Comment (bgavrilMS) [reply tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:149]: Fixed.

**Review by pmaytak — APPROVED** [review]

**Comment by bgavrilMS** [comment]
@pmaytak - no need to re-review, I pressed the button by mistake :)

**Review by pmaytak — APPROVED** [review]
Didn't review in much detail; skimmed for changes.

File: src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs [hunk 6]
```diff
@@ -136,15 +136,19 @@ internal class AuthorityInfo
 
         internal bool IsInstanceDiscoverySupported => AuthorityType == AuthorityType.Aad;
 
+        /// <summary>
+        /// For IWA
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/AppConfig/AuthorityInfo.cs:140]: Should this also be the case for ROPC?

File: tests/Microsoft.Identity.Test.Common/TestConstants.cs [hunk 7]
```diff
@@ -107,6 +108,13 @@ public static HashSet<string> s_scope
         public const string DstsAuthorityTenanted = "https://some.url.dsts.core.azure-test.net/dstsv2/" + TenantId + "/";
         public const string DstsAuthorityCommon = "https://some.url.dsts.core.azure-test.net/dstsv2/" + Common + "/";
 
+        public const string GenericAuthority = "https://demo.duendesoftware.com";
+
+        // not actually used by MSAL directly, MSAL will transform it to friendly name format
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Common/TestConstants.cs:113]: Only true for now. evetually ests will be updated to support all formats

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs [hunk 8]
```diff
@@ -62,25 +82,142 @@ public void WithTenantIdExceptions()
...
+            Assert.AreEqual(ex1.ErrorCode, MsalError.TenantOverrideNonAad);
+        }
+
+        [DataTestMethod]
+        [DataRow(TestConstants.DstsAuthorityCommon)]
+        [DataRow(TestConstants.DstsAuthorityTenanted)]
+        [DataRow(TestConstants.CiamAuthorityMainFormat)]
+        [DataRow(TestConstants.CiamAuthorityWithFriendlyName)]
+        [DataRow(TestConstants.CiamAuthorityWithGuid)]
+        public void WithTenantId_NonAad(string inputAuthority)
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:94]: nit: `Requestlevel_WithTenantId_NonAad`?

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs [hunk 9]
```diff
@@ -62,25 +82,142 @@ public void WithTenantIdExceptions()
...
+                (cca.AppConfig as ApplicationConfiguration).Authority.AuthorityInfo.Host,
+                "The host should have stayed the same");
+
+            Assert.AreEqual(
+                TestConstants.TenantId2, 
+                AuthorityHelpers.GetTenantId(new Uri(cca.Authority)));
+        }
+
+        [TestMethod]
+        public void B2CAuthorityWithTenant()
```
Comment (trwalke) [inline tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AuthorityTests.cs:147]: This name and the name of the test on 195 are a bit too similar, should propably add `AppLevel...` and `RequestLevel...` to differentiate. Maybe even group them.

**Review by trwalke — APPROVED** [review]

**Pull request merged**